src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/TestMarkupParser.cs

```diff
@@ -0,0 +1,406 @@
+                        }
+
+                        var textSpanBuilder = spansBuilder.GetOrAdd(startKey, _ => ImmutableArray.CreateBuilder<TextSpan>());
+                        textSpanBuilder.Add(TextSpan.FromBounds(startOutputPosition, endOutputPosition));
+
+                        startIndex--;
+                        continue;
+                    }
+                    else
+                    {
+                        startIndex++;
+                        continue;
+                    }
+                }
+            }
+
+            spans = spansBuilder.ToImmutableDictionary(pair => pair.Key, pair => pair.Value.ToImmutable());
+        }
+
+        /// <summary>
+        /// Parses the input markup to find standalone positions and the start and end positions of text spans.
+        /// </summary>
+        /// <param name="input">The input markup.</param>
+        /// <param name="output">The output content with markup syntax removed from <paramref name="input"/>.</param>
+        /// <param name="positions">A list of positions defined in markup (<c>$$</c>).</param>
+        /// <param name="startPositions">A list of starting positions of spans in markup. The key of the element is a
+        /// position (the location of the <c>[|</c> or <c>{|</c>). The value of the element is the <c>text</c> content
+        /// of a <c>{|text:</c> starting syntax, or <see langword="null"/> if the <c>[|</c> syntax was used. This list
+        /// preserves the original order of starting markup tags in the input.</param>
+        /// <param name="endPositions">A list of ending positions of spans in markup. The key of the element is a
+        /// position (the location of the <c>|]</c> or <c>|}</c>). The value of the element is the <c>#id</c> content of
+        /// a <c>|#id}</c> ending syntax, or <see langword="null"/> if the <c>|]</c> or <c>|}</c> syntax was used. This
+        /// list preserves the original order of the ending markup tags in the input.</param>
+        private void Parse(string input, out string output, out ImmutableArray<int> positions, out ImmutableArray<(int inputPosition, int outputPosition, string key)> startPositions, out ImmutableArray<(int inputPosition, int outputPosition, string key)> endPositions)
+        {
+            var positionsBuilder = ImmutableArray.CreateBuilder<int>();
+            var startPositionsBuilder = ImmutableArray.CreateBuilder<(int inputPosition, int outputPosition, string key)>();
+            var endPositionsBuilder = ImmutableArray.CreateBuilder<(int inputPosition, int outputPosition, string key)>();
+
+            var outputBuilder = new StringBuilder();
+
+            var currentIndexInInput = 0;
+            var inputOutputOffset = 0;
+
+            var matches = new List<(int position, string key)>(6);
+            while (true)
+            {
+                matches.Clear();
+
+                if (!_ignorePositionString)
```

Comment:
this is the only logic add.  everything else is extraction from one type to another.  i tried to just make this so we had one type, with an instance API and a static API.  However, C# doesn't allow the sigs to match (evne if one is instance and one is static).  so i had to pull into another.